balancer.Dialer panics "send on closed channel"

Encountered while running Lantern with my changes of detour. But this bug may not directly triggered by my changes. Worth take a look.

```
panic: send on closed channel

goroutine 8790 [running]:
github.com/getlantern/balancer.(*dialer).onError(0xc2081abca0, 0x4f3bcc0, 0xc20b7015b0)
	/Users/fffw/workspace/getlantern/lantern/src/github.com/getlantern/balancer/dialer.go:111 +0x6c
github.com/getlantern/balancer.(*Balancer).DialQOS(0xc208a2b980, 0x44f6130, 0x3, 0xc209e4c470, 0xd, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/fffw/workspace/getlantern/lantern/src/github.com/getlantern/balancer/balancer.go:95 +0x5e0
github.com/getlantern/balancer.(*Balancer).Dial(0xc208a2b980, 0x44f6130, 0x3, 0xc209e4c470, 0xd, 0x0, 0x0, 0x0, 0x0)
	/Users/fffw/workspace/getlantern/lantern/src/github.com/getlantern/balancer/balancer.go:105 +0x89
github.com/getlantern/balancer.*Balancer.Dial·fm(0x44f6130, 0x3, 0xc209e4c470, 0xd, 0x0, 0x0, 0x0, 0x0)
	/Users/fffw/workspace/getlantern/lantern/src/github.com/getlantern/flashlight/client/reverseproxy.go:47 +0x79
github.com/getlantern/detour.func·016()
	/Users/fffw/workspace/getlantern/lantern/src/github.com/getlantern/detour/detour_conn.go:23 +0x5e
created by github.com/getlantern/detour.DialDetour
	/Users/fffw/workspace/getlantern/lantern/src/github.com/getlantern/detour/detour_conn.go:30 +0x16e
```

--- a/src/github.com/getlantern/balancer/balancer.go
+++ b/src/github.com/getlantern/balancer/balancer.go
@@ -91,7 +91,7 @@ func (b *Balancer) DialQOS(network, addr string, targetQOS int) (net.Conn, error
 		}
 		conn, err := d.Dial(network, addr)
 		if err != nil {
-			log.Tracef("Unable to dial: %s", err)
+			log.Debugf("Unable to dial %s://%s: %s", network, addr, err)
 			d.onError(err)
 			continue
 		}
@@ -120,12 +120,12 @@ func randomDialer(dialers []*dialer, targetQOS int) (chosen *dialer, others []*d
 	filtered, highestQOS := dialersMeetingQOS(dialers, targetQOS)
 
 	if len(filtered) == 0 {
-		log.Trace("No dialers meet targetQOS, using those with highest QOS")
+		log.Tracef("No dialers meet targetQOS %d, using those with highestQOS %d", targetQOS, highestQOS)
 		filtered, _ = dialersMeetingQOS(dialers, highestQOS)
 	}
 
 	if len(filtered) == 0 {
-		log.Trace("Still no dialers!")
+		log.Debugf("No dialers meet targetQOS %d or highestQOS %d!", targetQOS, highestQOS)
 		return nil, nil
 	}
 
@@ -140,7 +140,7 @@ func randomDialer(dialers []*dialer, targetQOS int) (chosen *dialer, others []*d
 	for _, d := range filtered {
 		aw += d.Weight
 		if aw > t {
-			log.Trace("Reached random target value, using this dialer")
+			log.Tracef("Randomly selected dialer %s with weight %d, QOS %d", d.Label, d.Weight, d.QOS)
 			return d, withoutDialer(dialers, d)
 		}
 	}
@@ -160,7 +160,6 @@ func dialersMeetingQOS(dialers []*dialer, targetQOS int) ([]*dialer, int) {
 
 		highestQOS = d.QOS // don't need to compare since dialers are already sorted by QOS (ascending)
 		if d.QOS >= targetQOS {
-			log.Tracef("Including dialer with QOS %d meeting targetQOS %d", d.QOS, targetQOS)
 			filtered = append(filtered, d)
 		}
 	}

--- a/src/github.com/getlantern/balancer/balancer_test.go
+++ b/src/github.com/getlantern/balancer/balancer_test.go
@@ -232,7 +232,7 @@ func TestAll(t *testing.T) {
 	// Test failure
 	b = New(dialer4, dialer4)
 	_, err = b.Dial("tcp", addr)
-	assert.NoError(t, err, "Dialing should have failed")
+	assert.NoError(t, err, "Dialing should have succeeded as we have 2nd try")
 	assert.Equal(t, 2, atomic.LoadInt32(&d4attempts), "Wrong number of dial attempts on failed dialer")
 
 	// Test success after successful retest using default check

--- a/src/github.com/getlantern/balancer/balancer_test.go
+++ b/src/github.com/getlantern/balancer/balancer_test.go
@@ -97,6 +97,7 @@ func TestAll(t *testing.T) {
 		},
 	}
 	dialer2 := &Dialer{
+		Label:  "Dialer 2",
 		Weight: 10000000,
 		QOS:    1,
 		Dial: func(network, addr string) (net.Conn, error) {
@@ -106,6 +107,7 @@ func TestAll(t *testing.T) {
 	}
 	checkAttempts := int32(0)
 	dialer3 := &Dialer{
+		Label:  "Dialer 3",
 		Weight: 1,
 		QOS:    15,
 		Dial: func(network, addr string) (net.Conn, error) {
@@ -127,6 +129,7 @@ func TestAll(t *testing.T) {
 
 	d4attempts := int32(0)
 	dialer4 := &Dialer{
+		Label:  "Dialer 4",
 		Weight: 1,
 		QOS:    15,
 		Dial: func(network, addr string) (net.Conn, error) {
@@ -227,11 +230,9 @@ func TestAll(t *testing.T) {
 	}
 
 	// Test failure
-	b = New(dialer4)
+	b = New(dialer4, dialer4)
 	_, err = b.Dial("tcp", addr)
-	assert.Error(t, err, "Dialing should have failed")
-
-	time.Sleep(1 * time.Second)
+	assert.NoError(t, err, "Dialing should have failed")
 	assert.Equal(t, 2, atomic.LoadInt32(&d4attempts), "Wrong number of dial attempts on failed dialer")
 
 	// Test success after successful retest using default check

--- a/src/github.com/getlantern/balancer/balancer_test.go
+++ b/src/github.com/getlantern/balancer/balancer_test.go
@@ -272,43 +272,19 @@ var (
 )
 
 type failingConn struct {
-	conn         net.Conn
+	net.Conn
 	bytesRead    int
 	bytesWritten int
 }
 
 func (c *failingConn) Read(b []byte) (n int, err error) {
-	n, err = c.conn.Read(b[:c.bytesRead])
+	n, err = c.Conn.Read(b[:c.bytesRead])
 	err = failed
 	return
 }
 
 func (c *failingConn) Write(b []byte) (n int, err error) {
-	n, err = c.conn.Write(b[:c.bytesWritten])
+	n, err = c.Conn.Write(b[:c.bytesWritten])
 	err = failed
 	return
-}
-
-func (c *failingConn) LocalAddr() net.Addr {
-	return c.conn.LocalAddr()
-}
-
-func (c *failingConn) RemoteAddr() net.Addr {
-	return c.conn.RemoteAddr()
-}
-
-func (c *failingConn) SetDeadline(t time.Time) error {
-	return c.conn.SetDeadline(t)
-}
-
-func (c *failingConn) SetReadDeadline(t time.Time) error {
-	return c.conn.SetReadDeadline(t)
-}
-
-func (c *failingConn) SetWriteDeadline(t time.Time) error {
-	return c.conn.SetWriteDeadline(t)
-}
-
-func (c *failingConn) Close() error {
-	return c.conn.Close()
 }

--- a/src/github.com/getlantern/balancer/dialer.go
+++ b/src/github.com/getlantern/balancer/dialer.go
@@ -72,17 +72,17 @@ func (d *dialer) start() {
 		for {
 			if lastFailed.After(lastCheckSucceeded) {
 				atomic.StoreInt32(&d.active, 0)
-				log.Trace("Inactive, scheduling check")
+				log.Tracef("Mark dialer %s as inactive, scheduling check", d.Label)
 				timeout := time.Duration(consecCheckFailures*consecCheckFailures) * 100 * time.Millisecond
 				timer.Reset(timeout)
 			} else {
 				atomic.StoreInt32(&d.active, 1)
-				log.Trace("Active")
+				log.Tracef("Mark dialer %s as active", d.Label)
 			}
 			select {
 			case t, ok := <-d.errCh:
 				if !ok {
-					log.Trace("dialer stopped")
+					log.Tracef("Dialer %s stopped", d.Label)
 					if d.OnClose != nil {
 						d.OnClose()
 					}
@@ -128,10 +128,11 @@ func (d *dialer) defaultCheck() bool {
 	ok, timedOut, _ := withtimeout.Do(10*time.Second, func() (interface{}, error) {
 		resp, err := client.Get("http://www.google.com/humans.txt")
 		if err != nil {
-			log.Debugf("Error on testing humans.txt: %s", err)
+			log.Debugf("Error testing dialer %s to humans.txt: %s", d.Label, err)
 			return false, nil
 		}
 		resp.Body.Close()
+		log.Tracef("Tested dialer %s to humans.txt, status code %d", d.Label, resp.StatusCode)
 		return resp.StatusCode == 200, nil
 	})
 	return !timedOut && ok.(bool)

--- a/src/github.com/getlantern/balancer/dialer.go
+++ b/src/github.com/getlantern/balancer/dialer.go
@@ -52,12 +52,15 @@ var (
 
 type dialer struct {
 	*Dialer
-	active int32
-	errCh  chan time.Time
+	active  int32
+	closeCh chan interface{}
+	errCh   chan time.Time
 }
 
 func (d *dialer) start() {
 	d.active = 1
+	// to avoid blocking sender, make it buffered
+	d.closeCh = make(chan interface{}, 1)
 	d.errCh = make(chan time.Time, 1)
 	if d.Check == nil {
 		d.Check = d.defaultCheck
@@ -80,14 +83,13 @@ func (d *dialer) start() {
 				log.Tracef("Mark dialer %s as active", d.Label)
 			}
 			select {
-			case t, ok := <-d.errCh:
-				if !ok {
-					log.Tracef("Dialer %s stopped", d.Label)
-					if d.OnClose != nil {
-						d.OnClose()
-					}
-					return
+			case <-d.closeCh:
+				log.Tracef("Dialer %s stopped", d.Label)
+				if d.OnClose != nil {
+					d.OnClose()
 				}
+				return
+			case t := <-d.errCh:
 				lastFailed = t
 			case <-timer.C:
 				ok := d.Check()
@@ -116,7 +118,7 @@ func (d *dialer) onError(err error) {
 }
 
 func (d *dialer) stop() {
-	close(d.errCh)
+	d.closeCh <- nil
 }
 
 func (d *dialer) defaultCheck() bool {

